src/modules/formBuilder/components/FormBuilder.tsx

```diff
@@ -0,0 +1,224 @@
+import { FetchResult } from '@apollo/client';
+import { LoadingButton } from '@mui/lab';
+import { Button, Paper, Typography } from '@mui/material';
+import { Box, Stack } from '@mui/system';
+import { isEqual } from 'lodash-es';
+import {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useMemo,
+  useState,
+} from 'react';
+import ConfirmationDialog from '@/components/elements/ConfirmationDialog';
+import Loading from '@/components/elements/Loading';
+import ErrorAlert from '@/modules/errors/components/ErrorAlert';
+import { ErrorState } from '@/modules/errors/util';
+import { updateFormItem } from '@/modules/form/util/formUtil';
+import FormBuilderHeader from '@/modules/formBuilder/components/FormBuilderHeader';
+import FormBuilderPalette from '@/modules/formBuilder/components/FormBuilderPalette';
+import FormItemEditor from '@/modules/formBuilder/components/FormItemEditor';
+import FormTree from '@/modules/formBuilder/components/formTree/FormTree';
+import {
+  DisabledDisplay,
+  EnableBehavior,
+  FormDefinitionFieldsForEditorFragment,
+  FormDefinitionJson,
+  FormItem,
+  UpdateFormDefinitionMutation,
+} from '@/types/gqlTypes';
+
+interface FormBuilderProps {
+  formDefinition: FormDefinitionFieldsForEditorFragment;
+  workingDefinition?: FormDefinitionJson;
+  setWorkingDefinition?: Dispatch<
+    SetStateAction<FormDefinitionJson | undefined>
+  >;
+  errorState?: ErrorState;
+  onSave: (
+    formDefinition: FormDefinitionJson
+  ) => Promise<FetchResult<UpdateFormDefinitionMutation>>;
+  saveLoading: boolean;
+  lastUpdatedDate?: string;
+  lastUpdatedBy?: string;
+  selectedItem?: FormItem;
+  setSelectedItem: Dispatch<SetStateAction<FormItem | undefined>>;
+  originalLinkId?: string;
+  setOriginalLinkId: Dispatch<SetStateAction<string | undefined>>;
+  closeItemEditor: () => void;
+}
+
+const FormBuilder: React.FC<FormBuilderProps> = ({
+  formDefinition,
+  workingDefinition,
+  setWorkingDefinition,
+  errorState,
+  onSave,
+  saveLoading,
+  lastUpdatedDate,
+  lastUpdatedBy,
+  selectedItem,
+  setSelectedItem,
+  originalLinkId,
+  setOriginalLinkId,
+  closeItemEditor,
+}) => {
+  const dirty = useMemo(() => {
+    return !isEqual(workingDefinition, formDefinition.definition);
+  }, [workingDefinition, formDefinition.definition]);
+
+  const [isBlocked, setIsBlocked] = useState(false);
+  const [unblockCallback, setUnblockCallback] = useState<
```

Comment:
Can you add some comments explaining these states? (I think I get it now, but it took me a bit).

Reply:
It also doesn't seem like we really need 2 pieces of state for this. We always set them together and clear them together, it seems. Maybe it would be simpler to have a single piece of state `blockedActionFunction` that is either opening the panel, navigating to the preview, or undefined.